catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,434 @@
+  private final ListeningExecutorService sourceAvailabilityChecksThreadPool;
+
+  private ScheduledExecutorService pollAllSourcesThreadPool;
+
+  private final Cache<SourceKey, SourceAvailability> cache;
+
+  public SourcePoller(
+      ExecutorService sourceAvailabilityChecksThreadPool,
+      ScheduledExecutorService pollAllSourcesThreadPool) {
+    this(
+        sourceAvailabilityChecksThreadPool,
+        pollAllSourcesThreadPool,
+        DEFAULT_POLL_INTERVAL,
+        DEFAULT_POLL_INTERVAL_TIME_UNIT);
+  }
+
+  /** Starts a process to periodically call {@link #pollAllSources()} */
+  @VisibleForTesting
+  SourcePoller(
+      ExecutorService sourceAvailabilityChecksThreadPool,
+      ScheduledExecutorService pollAllSourcesThreadPool,
+      final long pollInterval,
+      final TimeUnit pollIntervalTimeUnit) {
+    this.sourceAvailabilityChecksThreadPool =
+        MoreExecutors.listeningDecorator(sourceAvailabilityChecksThreadPool);
+    this.pollAllSourcesThreadPool = pollAllSourcesThreadPool;
+
+    LOGGER.debug(
+        "Scheduling ExecutorService at fixed rate of {} {} with an initial delay of {}",
+        pollInterval,
+        pollIntervalTimeUnit,
+        pollInterval);
+    pollAllSourcesThreadPool.scheduleAtFixedRate(
+        this::pollAllSources, pollInterval, pollInterval, pollIntervalTimeUnit);
+
+    cache =
+        CacheBuilder.newBuilder()
+            .maximumSize(1000)
+            .removalListener(
+                (RemovalNotification<SourceKey, SourceAvailability> removalNotification) -> {
+                  LOGGER.debug(
+                      "Removing SourceKey {} from cache. cause={}. Cancelling any futures that are checking availability for that entry",
+                      removalNotification.getKey(),
+                      removalNotification.getCause());
+                })
+            .build();
+
+    setPollInterval(pollInterval, pollIntervalTimeUnit);
+  }
 
   /**
-   * Constructor to schedule the SourcePollerRunner to execute immediately and at a fixed interval,
-   * currently set at every 1 minute. This constructor is invoked by the CatalogFramework's
-   * blueprint.
+   * This method is used to keep the cache up to date.
    *
-   * @param incomingRunner the SourcePollerRunner to use for polling
+   * <p>The current {@link Source}s configured by {@link #setConnectedSources(List)}, {@link
+   * #setFederatedSources(List)}, {@link #setCatalogProviders(List)}, and {@link
+   * #setCatalogStores(List)} are discovered.
+   *
+   * <p>A process to check (or recheck) and cache the availability of each {@link Source} is
+   * performed.
    */
-  public SourcePoller(SourcePollerRunner incomingRunner) {
+  public synchronized void pollAllSources() {
+    try {
+      final Set<Source> sources =
+          Stream.of(connectedSources, federatedSources, catalogProviders, catalogStores)
+              .flatMap(Collection::stream)
+              .collect(Collectors.toSet());
```

Comment:
❓I'm not sure here of the value of collecting to a `Set` instead of a `List`. If those sources do not implement `hashCode` and `equals`, then this doesn't do any good, and if they do and do it wrong (it's really difficult to have good `equals` implementation that respect the `equals` contract when you're dealing with different sub-classes), we could end up with two sources stepping on each other or get some unexpected behaviors.